server/routerlicious/packages/services-client/src/ipUtils.ts

```diff
@@ -0,0 +1,99 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/* eslint-disable no-bitwise */
+/* eslint-disable max-len */
+import { IPv6, IPv4 } from "ipaddr.js";
+
+/**
+ * The network information from the client IP address
+ *
+ * @internal
+ */
+export class NetworkInformation {
+	public isPrivateLink!: boolean;
+	public privateLinkId?: string;
+	public privateIpAddress?: string;
+}
+
+/*
+    The source IPv6 source address carries vnet and private link information as below:
+
+    | byte 1     | byte 2                               | bytes 3-6                | bytes 7-10         | bytes 11-12  | bytes 13-16         |
+    |            |       |            |        |        |                          | VNetTrafficTag =   |              |                     |
+    | ULA prefix |1: V1  |0:ST        |Reserved|ST      | SNAT VIP OR Pvt Link ID  | Region Id + VNet Id|Subnet Id     |IPv4 Customer address|
+    | 8 bits     |0: V2  |1:Pvt Link  |5 bit   |Policy  |  4 bytes = 32 bits       | 4 bytes = 32 bits  |16 bits       |4 bytes = 32 bits    |
+    |NSM reserved|9th bit|10th bit    |        |16th bit| Used for encoding Pvt Lnk| Region Id : 8 bit  |              |Used for billing     |
+    |            |       |            |        |        | ID OR SNAT VIP of service| Vnet Id : 24 bit   |              |                     |
+*/
+const PrivateLinkIpStructure = {
+	packetVersionBit: 9,
+	isPrivateLinkBit: 10,
+	privateLinkIdMostSignificantByte: 6,
+	privateAddressMostSignificantByte: 13,
+};
+
+/**
+ * Get the network information from the client IP address, including whether it is a private link and the private link ID.
+ *
+ * @internal
+ */
+export function getNetworkInformationFromIP(clientIp?: string): NetworkInformation {
```

Comment:
This function is pretty complex and involves byte operation. Is it borrowed from somewhere? Should we add unit test for it?

Reply:
Yup. :) I just borrow it from here:
https://msazure.visualstudio.com/One/_git/Mgmt-AppInsights-DevExp-API?path=%2FDraft%2FDraft.Role%2Fsrc%2Futils%2FipUtils.ts&_a=contents&version=GBmaster